### PR TITLE
Fix highllighter, implements hotkeys

### PR DESCRIPTION
**Pull request by opastushkov**

### CODE REVIEW

File: src/editor/codeeditor.cpp [hunk 1]
```diff
@@ -105,6 +109,8 @@ CodeEditor::~CodeEditor()
 
 void CodeEditor::setTextColors()
 {
+    fmtUndefined.setUnderlineStyle(QTextCharFormat::WaveUnderline);
+    fmtUndefined.setUnderlineColor(Qt::red);
```
Comment (PeterLysyk) [inline src/editor/codeeditor.cpp]: create config parametr with this color. Do the same as for "ftm" variables belowe. You can set this value in the configparamers.h file

File: src/editor/codeeditor.cpp [hunk 2]
```diff
@@ -142,6 +148,13 @@ void CodeEditor::setConfigParam(const ConfigParams &configParam)
     mConfigParam = configParam;
 }
 
+void CodeEditor::handleLinesSwap(int firstLine, int secondLine)
```
Comment (PeterLysyk) [inline src/editor/codeeditor.cpp]: pass const parametrs
`(const int firstLine, const int secondLine)`

File: src/editor/codeeditor.cpp [hunk 3]
```diff
@@ -190,6 +205,11 @@ void CodeEditor::handleLineChange(int lastLineWithChange)
     int lineDifference = currentLinesCount - mLinesCount;
     mLinesCount = currentLinesCount;
 
+    if(!mLcpp->wasRunning)
```
Comment (PeterLysyk) [inline src/editor/codeeditor.cpp]: is it variable? rename or create the same name method

Comment (PeterLysyk) [inline src/editor/codeeditor.cpp]: if (!mLcpp->wasRunning)

File: src/editor/codeeditor.cpp [hunk 4]
```diff
@@ -380,7 +393,11 @@ void CodeEditor::setZoom(int zoomVal)
 
 void CodeEditor::textChangedInTheOneLine()
 {
-    emit(textChangedInLine(this->textCursor().blockNumber()));
+    if(mCode != document()->toPlainText())
```
Comment (PeterLysyk) [inline src/editor/codeeditor.cpp]: if (mCode != document()->toPlainText())

File: src/editor/codeeditor.cpp [hunk 5]
```diff
@@ -740,17 +757,30 @@ void formating(QTextCharFormat fmt, QTextCursor &cursor, Token token, int starti
 void CodeEditor::highlightText()
 {
     // Set cursor position to begining of visible area
-    QTextCursor cursor = this->cursorForPosition(QPoint(0, 0));
-    int firstVisibleLine = cursor.blockNumber();
+//    QTextCursor cursor = this->cursorForPosition(QPoint(0, 0));
```
Comment (PeterLysyk) [inline src/editor/codeeditor.cpp]: don't leave commented code

File: src/editor/codeeditor.cpp [hunk 6]
```diff
@@ -740,17 +757,30 @@ void formating(QTextCharFormat fmt, QTextCursor &cursor, Token token, int starti
...
+//    qDebug() << startingPosition;
+
     // Set cursor to end of visible aread
     QPoint bottom_right(this->viewport()->width() - 1, this->viewport()->height() - 1);
     cursor = this->cursorForPosition(bottom_right);
     int lastVisibleLine = cursor.blockNumber();
 
     // Highlight visible area
-    for(int i = firstVisibleLine; i <= lastVisibleLine; ++i)
+    for(int i = start; i <= lastVisibleLine; ++i)
```
Comment (PeterLysyk) [inline src/editor/codeeditor.cpp]: 1) for (int i = start; i <= lastVisibleLine; ++i)
2) better use **auto** . _i_ can't be negative, so you don't need to create type as int

File: src/editor/keypressevents.cpp [hunk 7]
```diff
@@ -146,6 +146,61 @@ void EventSendLexem::operator()(CodeEditor * codeEditor, QKeyEvent *e)
...
+QString getLineUnderCursor(QTextCursor &cursor)
+{
+    cursor.select(QTextCursor::LineUnderCursor);
+    return cursor.selectedText();
+}
+
+void EventCtrlUpArrow::operator()(CodeEditor *codeEditor, QKeyEvent *e)
+{
+    QTextCursor cursor = codeEditor->textCursor();
+    if(cursor.blockNumber() > 0)
```
Comment (PeterLysyk) [inline src/editor/keypressevents.cpp]: if (cursor.blockNumber() > 0)

File: src/editor/keypressevents.cpp [hunk 8]
```diff
@@ -146,6 +146,61 @@ void EventSendLexem::operator()(CodeEditor * codeEditor, QKeyEvent *e)
...
+        codeEditor->setTextCursor(cursor);
+    }
+}
+EventCtrlUpArrow::~EventCtrlUpArrow() = default;
+
+void EventCtrlDownArrow::operator()(CodeEditor *codeEditor, QKeyEvent *e)
+{
+    QTextCursor cursor = codeEditor->textCursor();
+
+    if(cursor.blockNumber() < codeEditor->document()->blockCount() - 1)
```
Comment (PeterLysyk) [inline src/editor/keypressevents.cpp]: if (cursor.blockNumber() < codeEditor->document()->blockCount() - 1)

File: src/editor/lexercpp.cpp [hunk 9]
```diff
@@ -269,6 +269,7 @@ void LexerCPP::handleLiteralState(const QChar &sym)
 
 void LexerCPP::lexicalAnalysis(QString code)
 {
+    wasRunning = true;
```
Comment (PeterLysyk) [inline src/editor/lexercpp.cpp]: great variable name!

File: src/editor/lexercpp.h [hunk 10]
```diff
@@ -13,9 +13,9 @@
 class LexerCPP: public iLexer
 {
 private:
-    QVector<QVector<Token>> mTokens;
     QVector<Token> mTokensOnCurrentLine;
     State mState;
+    bool isCommentBlock;
```
Comment (PeterLysyk) [inline src/editor/lexercpp.h:19]: 1)do you use this variable?
2)mIsCommentBlock.
3) Why is it variable?
4) Why do you named variable as a method?

File: src/editor/token.h [hunk 11]
```diff
@@ -12,6 +12,13 @@ struct Token
     int mEnd;
     Token(QString name, State type, unsigned int begin, unsigned int end):
           mName(name), mType(type), mBegin(begin), mEnd(end) {}
+    bool operator== (Token argument)
```
Comment (PeterLysyk) [inline src/editor/token.h]: const Token argument

**Review by PeterLysyk — APPROVED** [review]
consider this remarks, please!

**Review by vsakharevych — APPROVED** [review]

**Pull request merged**